File: nobodywho/kotlin/common/src/Tts.kt
package ai.nobodywho

import java.io.Closeable
import uniffi.nobodywho.RustTts as InternalTts
import uniffi.nobodywho.loadTts as loadInternalTts

enum class TtsArchitecture(internal val value: String) {
    KOKORO("kokoro"),
    SUPERTONIC("supertonic")
}

enum class TtsDevice(internal val value: String) {
    AUTO("auto"),
    CPU("cpu"),
    CUDA("cuda")
}

/** Text-to-speech synthesizer that returns WAV bytes. */
class Tts private constructor(
    private val inner: InternalTts
) : Closeable {
    /** Create a TTS synthesizer synchronously. */
    constructor(
        source: String,
        architecture: TtsArchitecture? = null,
        voice: String? = null,
        language: String? = null,
        speed: Float? = null,
        steps: UInt? = null,
        silenceDuration: Float? = null,
        device: TtsDevice = TtsDevice.AUTO
    ) : this(
        InternalTts(
            source = source,
            architecture = architecture?.value,
            voice = voice,
            language = language,
            speed = speed,
            steps = steps,
            silenceDuration = silenceDuration,
            device = device.value
        )
    )

    companion object {
        /** Create a TTS synthesizer asynchronously. */
        suspend fun load(
            source: String,
            architecture: TtsArchitecture? = null,
            voice: String? = null,
            language: String? = null,
            speed: Float? = null,
            steps: UInt? = null,
            silenceDuration: Float? = null,
            device: TtsDevice = TtsDevice.AUTO
        ): Tts = Tts(
            loadInternalTts(
                source = source,
                architecture = architecture?.value,
                voice = voice,
                language = language,
                speed = speed,
                steps = steps,
                silenceDuration = silenceDuration,
                device = device.value
            )
        )
    }

    /** Synthesize text and return WAV bytes. */
    suspend fun synthesize(text: String): ByteArray = inner.synthesizeAsync(text)

    /** Synthesize text synchronously and return WAV bytes. */
    fun synthesizeSync(text: String): ByteArray = inner.synthesize(text)

    /** Free the underlying Rust resources. */
    fun destroy() = inner.destroy()
    override fun close() { destroy() }
}

File: nobodywho/swift/src/Tts.swift
import Foundation
import NobodyWhoGenerated

public enum TtsArchitecture: String {
    case kokoro
    case pocketTts = "pocket-tts"
    case supertonic
}

public enum TtsDevice: String {
    case auto
    case cpu
    case cuda
}

/// Text-to-speech synthesizer that returns WAV bytes.
public class Tts {
    private let inner: NobodyWhoGenerated.RustTts

    /// Create a TTS synthesizer synchronously.
    public init(
        source: String,
        architecture: TtsArchitecture? = nil,
        voice: String? = nil,
        language: String? = nil,
        speed: Float? = nil,
        steps: UInt32? = nil,
        silenceDuration: Float? = nil,
        precision: String? = nil,
        temperature: Float? = nil,
        huggingfaceToken: String? = nil,
        device: TtsDevice = .auto
    ) throws {
        self.inner = try NobodyWhoGenerated.RustTts(
            source: source,
            architecture: architecture?.rawValue,
            voice: voice,
            language: language,
            speed: speed,
            steps: steps,
            silenceDuration: silenceDuration,
            precision: precision,
            temperature: temperature,
            huggingfaceToken: huggingfaceToken,
            device: device.rawValue
        )
    }

    private init(inner: NobodyWhoGenerated.RustTts) {
        self.inner = inner
    }

    /// Create a TTS synthesizer asynchronously.
    public static func load(
        source: String,
        architecture: TtsArchitecture? = nil,
        voice: String? = nil,
        language: String? = nil,
        speed: Float? = nil,
        steps: UInt32? = nil,
        silenceDuration: Float? = nil,
        precision: String? = nil,
        temperature: Float? = nil,
        huggingfaceToken: String? = nil,
        device: TtsDevice = .auto
    ) async throws -> Tts {
        let inner = try await NobodyWhoGenerated.loadTts(
            source: source,
            architecture: architecture?.rawValue,
            voice: voice,
            language: language,
            speed: speed,
            steps: steps,
            silenceDuration: silenceDuration,
            precision: precision,
            temperature: temperature,
            huggingfaceToken: huggingfaceToken,
            device: device.rawValue
        )
        return Tts(inner: inner)
    }

    /// Synthesize text and return WAV bytes.
    public func synthesize(_ text: String) async throws -> Data {
        try await inner.synthesizeAsync(text: text)
    }

    /// Synthesize text synchronously and return WAV bytes.
    public func synthesizeSync(_ text: String) throws -> Data {
        try inner.synthesize(text: text)
    }
}
